src/v/net/tls_certificate_probe.h

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "bytes/bytes.h"
+#include "config/tls_config.h"
+#include "seastarx.h"
+#include "ssx/metrics.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/net/tls.hh>
+
+#include <chrono>
+#include <exception>
+#include <iosfwd>
+
+namespace net {
+
+class tls_certificate_probe {
+    using tls_serial_number
+      = named_type<uint32_t, struct tls_serial_number_tag>;
+
+public:
+    using clock_type = ss::lowres_system_clock;
+    tls_certificate_probe() { reset(); }
+    tls_certificate_probe(const tls_certificate_probe&) = delete;
+    tls_certificate_probe& operator=(const tls_certificate_probe&) = delete;
+    tls_certificate_probe(tls_certificate_probe&&) = delete;
+    tls_certificate_probe& operator=(tls_certificate_probe&&) = delete;
+    ~tls_certificate_probe() = default;
+
+    void loaded(
+      const ss::tls::certificate_credentials& creds, std::exception_ptr ex);
+
+    void setup_metrics(std::string_view area, std::string_view detail);
+
+private:
+    ssx::metrics::all_metrics_groups _metrics;
+    std::chrono::seconds _load_time;
+    std::chrono::seconds _cert_expiry_time;
+    std::chrono::seconds _ca_expiry_time;
+    tls_serial_number _cert_serial;
+    tls_serial_number _ca_serial;
+    bool _cert_loaded{false};
+
+    bool cert_valid() const {
+        auto now = std::chrono::duration_cast<std::chrono::seconds>(
+          clock_type::now().time_since_epoch());
+        return _cert_loaded && now < _ca_expiry_time && now < _cert_expiry_time;
+    }
+
+    void reset() {
+        _cert_loaded = false;
+        _cert_expiry_time = std::chrono::seconds::max();
+        _cert_serial = {};
+        _ca_expiry_time = std::chrono::seconds::max();
+        _ca_serial = {};
+    }
+
+    friend std::ostream&
+    operator<<(std::ostream& o, const tls_certificate_probe& p);
+};
+
+ss::future<ss::shared_ptr<ss::tls::server_credentials>>
+build_reloadable_server_credentials_with_probe(
+  config::tls_config tls_config,
+  std::string_view service,
+  std::string_view listener_name,
+  ss::tls::reload_callback cb = {});
+
+template<typename T>
+concept TLSCreds = std::is_base_of<ss::tls::certificate_credentials, T>::value;
+
+template<TLSCreds T>
```

Comment:
to keep the header clean you can you still move the implementation here into the .cc file and explicitly instantiate the template for cases that exist. maybe that's not tractable in general, but it looks like this might only have 2 types used?